File: C++/0003-longest-substring-without-repeating-characters.cpp
// Author   : Aniruddha Krishna Jha
// Date     : 13/08/2022
// Link     : https://leetcode.com/problems/longest-substring-without-repeating-characters/

/*******************************************************************************
Given a string s, find the length of the longest substring without repeating characters.

Example 1:

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Example 2:

Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
*******************************************************************************/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        //create a set of char, window init
        unordered_set<char>st;
        int l = 0, r = 0, mlen = 0;
        
        while(r < s.size()){
            //if char not found, insert in set extend the window to right 
            if(st.count(s[r]) == false){
                st.insert(s[r++]);
                mlen = max(mlen, r - l);
            }
            else{
                st.erase(s[l++]);               
            }

            //else, remove from set shorten the window from left
        }
        return mlen;
    }
};

File: Python/0003-longest-substring-without-repeating-characters.py
# Author    : Aniruddha Krishna Jha   
# Date      : 02/07/2023
# Link      : https://leetcode.com/problems/longest-substring-without-repeating-characters/

'''********************************************************************************** 
Given a string s, find the length of the longest  substring without repeating characters.

Example 1:

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Example 2:

Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
**********************************************************************************'''

class Solution:
    '''
    Sliding Window -> keep members in Hashset 
    Update max length after new addition

    Time Complexity: O(n)
    Space Complexity: O(n)
    '''
    def lengthOfLongestSubstring(self, s: str) -> int:
        l, res = 0, 0
        hashSet = set()

        for r in range(len(s)):
            # if already in set, create new window from this pt as new end
            while s[r] in hashSet:
                hashSet.remove(s[l])
                l += 1
            hashSet.add(s[r])
            res = max(res, r - l + 1)
        return res